src/cli/self_update/windows.rs

```diff
@@ -212,63 +212,72 @@ fn get_windows_path_var() -> Result<Option<String>> {
             if let Some(s) = utils::string_from_winreg_value(&val) {
                 Ok(Some(s))
             } else {
-                warn!("the registry key HKEY_CURRENT_USER\\Environment\\PATH does not contain valid Unicode. \
-                       Not modifying the PATH variable");
+                warn!(
+                    "the registry key HKEY_CURRENT_USER\\Environment\\PATH is not a string. \
+                       Not modifying the PATH variable"
+                );
                 Ok(None)
             }
         }
-        Err(ref e) if e.kind() == io::ErrorKind::NotFound => Ok(Some(String::new())),
+        Err(ref e) if e.kind() == io::ErrorKind::NotFound => Ok(Some(Vec::new())),
         Err(e) => Err(e).chain_err(|| ErrorKind::WindowsUninstallMadness),
     }
 }
 
 // Returns None if the existing old_path does not need changing, otherwise
 // prepends the path_str to old_path, handling empty old_path appropriately.
-fn _add_to_path(old_path: &str, path_str: String) -> Option<String> {
+fn _add_to_path(old_path: Vec<u16>, path_str: Vec<u16>) -> Option<Vec<u16>> {
     if old_path.is_empty() {
         Some(path_str)
-    } else if old_path.contains(&path_str) {
+    } else if old_path
+        .windows(path_str.len())
+        .any(|path| path == path_str)
+    {
         None
     } else {
         let mut new_path = path_str;
-        new_path.push_str(";");
-        new_path.push_str(&old_path);
+        new_path.push(b';' as u16);
+        new_path.extend_from_slice(&old_path);
         Some(new_path)
     }
 }
 
 // Returns None if the existing old_path does not need changing
-fn _remove_from_path(old_path: &str, path_str: String) -> Option<String> {
-    let idx = old_path.find(&path_str)?;
+fn _remove_from_path(old_path: Vec<u16>, path_str: Vec<u16>) -> Option<Vec<u16>> {
+    let idx = old_path
+        .windows(path_str.len())
+        .position(|path| path == path_str)?;
```

Comment:
@rbtcollins

This function can't be implemented with only `OsString`s. This is why I used `Vec<u16>` everywhere.

Should I rewrite it to take and return `OsString`s and convert to `Vec<u16>` and back ? Or can I just leave `Vec<u16>` everywhere because that's what can be edited ?

Reply:
So _remove_from_path is the only complicated case right? Everywhere else is simple comparison or push or clear cases. I think constraining the semantics to signal that OsString is the intent, and containing the complexity to this one place is quite tolerable - it won't be a performance problem after all. 

OTOH the Vec<u16> is already contained in this file, so it won't spread. Ultimately I'm going to be ok either way, and I'm mainly happy to have this contribution.

But my preference, if you're up for it, is to use OsString most places.

Reply:
Yes, it is the only one, but it is also half of them (the other one being `_add_to_path`). Note that it will also require converting between `OsString` and `Vec` in `string_from_winreg_value` and `string_to_winreg_bytes`. Since this file is windows-specific (and to avoid four type conversions) I'd rather keep `Vec`s.

That said now I've explained why I did things the way I did I'll be fine with whatever your decision is.

Reply:
In fact, we also need `Vec`s in `_add_to_path` to check whether cargo_home is already in the path, which makes me quite strongly against converting `OsString`s back and forth for no real benefit. (Though I could add `type OsString = Vec<u16>` and then write `OsString` instead of `Vec<u16>` is you really want)